# Data Preparation and Cleaning
In this notebook I'll clean the dataset obtained from kaggle and save it so I can perform analysis.

## Loading libraries

In [152]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
import seaborn as sns 

## Loading the dataset
### Lead interaction Details 

In [153]:
# Load the interaction details 
interactions_df = pd.read_csv('leads_interaction_details.csv')
interactions_df.head(7)

jnr_sm_id  lead_id     lead_stage call_done_date call_status  \
0  JNR1001MG  USR1001           lead       1/2/2022  successful   
1  JNR1001MG  USR1001           lead       1/2/2022  successful   
2  JNR1001MG  USR1002           lead       1/3/2022  successful   
3  JNR1001MG  USR1002           lead       1/4/2022  successful   
4  JNR1001MG  USR1002      awareness       1/5/2022  successful   
5  JNR1001MG  USR1002      awareness       1/6/2022  successful   
6  JNR1001MG  USR1002  consideration       1/7/2022  successful   

                  call_reason  
0           lead_introduction  
1               demo_schedule  
2           lead_introduction  
3               demo_schedule  
4          post_demo_followup  
5  followup_for_consideration  
6   interested_for_conversion

In [154]:
interactions_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2192 entries, 0 to 2191
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   jnr_sm_id       2192 non-null   object
 1   lead_id         2192 non-null   object
 2   lead_stage      2192 non-null   object
 3   call_done_date  2192 non-null   object
 4   call_status     2192 non-null   object
 5   call_reason     2192 non-null   object
dtypes: object(6)
memory usage: 102.9+ KB


The dataset has 2192 rows and 5 columns. There are no missing values.

In [155]:
interactions_df.duplicated().any()

True

The dataset has duplicated rows. 

In [156]:
interactions_df.describe()

jnr_sm_id  lead_id lead_stage call_done_date call_status  \
count        2192     2192       2192           2192        2192   
unique         16      358          4             40           2   
top     JNR1012MG  USR1193       lead       1/8/2022  successful   
freq          164       11       1547            156        1866   

              call_reason  
count                2192  
unique                  8  
top     demo_not_attended  
freq                  701

### Client basic details

In [157]:
# Load the client details
leads_details = pd.read_csv('leads_basic_details.csv')
leads_details.sample(10)

lead_id  age  gender   current_city current_education  \
32   USR1033   18  FEMALE      Bengaluru            B.Tech   
134  USR1135   20    MALE  Visakhapatnam            B.Tech   
326  USR1327   24  FEMALE      Hyderabad   Looking for Job   
46   USR1047   18  FEMALE        Chennai            B.Tech   
345  USR1346   25  FEMALE         Mumbai   Looking for Job   
299  USR1300   20  FEMALE  Visakhapatnam            B.Tech   
315  USR1316   18    MALE        Chennai            B.Tech   
54   USR1055   24  FEMALE  Visakhapatnam   Looking for Job   
174  USR1175   18    MALE      Bengaluru            B.Tech   
263  USR1264   22  FEMALE      Hyderabad   Looking for Job   

       parent_occupation  lead_gen_source  
32   Government Employee     social_media  
134  Government Employee  email_marketing  
326               Lawyer              SEO  
46           IT Employee  email_marketing  
345               Lawyer              SEO  
299               Lawyer   user_referrals  
315             Business              SEO  
54                Lawyer     social_media  
174          IT Employee   user_referrals  
263             Business     social_media

In [158]:
leads_details.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 360 entries, 0 to 359
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   lead_id            360 non-null    object
 1   age                360 non-null    int64 
 2   gender             360 non-null    object
 3   current_city       360 non-null    object
 4   current_education  360 non-null    object
 5   parent_occupation  360 non-null    object
 6   lead_gen_source    360 non-null    object
dtypes: int64(1), object(6)
memory usage: 19.8+ KB


The datset contains details of 360 clients with information like age, gender, current city and education, parent's occupation and how the client came to know about the product.

In [159]:
leads_details.duplicated().any()

False

There are no duplicated rows. 

In [160]:
leads_details.describe()

age
count  360.000000
mean    21.561111
std     11.555444
min     16.000000
25%     18.000000
50%     21.000000
75%     24.000000
max    211.000000

There are outliers in this dataset.

In [161]:
demos_df = pd.read_csv('leads_demo_watched_details.csv')
demos_df.sample(10)

lead_id demo_watched_date language  watched_percentage
150  USR1247         1/10/2022   Telugu                  76
188  USR1315         1/18/2022  English                  44
189  USR1317         2/25/2022  English                  48
133  USR1213         1/20/2022   Telugu                 233
193  USR1348         2/27/2022  English                  72
146  USR1237         2/21/2022  English                  80
61   USR1075         1/18/2022  English                  44
5    USR1007         1/11/2022  English                  52
132  USR1212         1/26/2022  English                  19
75   USR1103          1/1/2022   Telugu                  41

In [162]:
demos_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 194 entries, 0 to 193
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   lead_id             194 non-null    object
 1   demo_watched_date   194 non-null    object
 2   language            194 non-null    object
 3   watched_percentage  194 non-null    int64 
dtypes: int64(1), object(3)
memory usage: 6.2+ KB


The dataset contains details of demos watched by leads. There are no missing values. 

In [163]:
demos_df.describe()

watched_percentage
count          194.000000
mean            56.634021
std             43.555635
min              2.000000
25%             35.000000
50%             55.500000
75%             75.750000
max            510.000000

There are leads who watched 510% of the demo. This is ofcourse is an outlier so it should be cleaned.

### Reasons for not being interested
This dataset contains the leads' reasons as to why they dropped out at the various stages of the customer acquistion process.

In [164]:
no_interest_df = pd.read_csv('leads_reasons_for_no_interest.csv')
no_interest_df

lead_id reasons_for_not_interested_in_demo  \
0    USR1001                No time for student   
1    USR1003                                NaN   
2    USR1004                                NaN   
3    USR1005                                NaN   
4    USR1006                                NaN   
..       ...                                ...   
289  USR1356                      Cannot afford   
290  USR1357                      Cannot afford   
291  USR1358              Wants offline classes   
292  USR1359            Will join in final year   
293  USR1360            Will join in final year   

    reasons_for_not_interested_to_consider  \
0                                      NaN   
1                      No time for student   
2                    Wants offline classes   
3                             Can't afford   
4         Student not interested in domain   
..                                     ...   
289                                    NaN   
290                                    NaN   
291                                    NaN   
292                                    NaN   
293                                    NaN   

    reasons_for_not_interested_to_convert  
0                                     NaN  
1                                     NaN  
2                                     NaN  
3                                     NaN  
4                                     NaN  
..                                    ...  
289                                   NaN  
290                                   NaN  
291                                   NaN  
292                                   NaN  
293                                   NaN  

[294 rows x 4 columns]

In [165]:
no_interest_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294 entries, 0 to 293
Data columns (total 4 columns):
 #   Column                                  Non-Null Count  Dtype 
---  ------                                  --------------  ----- 
 0   lead_id                                 294 non-null    object
 1   reasons_for_not_interested_in_demo      164 non-null    object
 2   reasons_for_not_interested_to_consider  79 non-null     object
 3   reasons_for_not_interested_to_convert   51 non-null     object
dtypes: object(4)
memory usage: 9.3+ KB


We have missing values in the last 3 columns. We can melt them to 2 columns, which stage they dropped out and the reason.

In [166]:
for col in no_interest_df.columns[1:]:
    print(no_interest_df[col].unique())

['No time for student' nan 'Student not interested in domain'
 'Wants offline classes' "Can't afford" 'Cannot afford'
 'Will join in final year']
[nan 'No time for student' 'Wants offline classes' "Can't afford"
 'Student not interested in domain' 'Will join in final year']
[nan "Can't afford" 'Will join in final year' 'No time for student'
 'Student not interested in domain' 'Wants offline classes']


Reasons like `Can't afford` and `Cannot afford` are the same, so it can be combined into one.

## Cleaning the data and wrangling it

In [167]:
# Making copies of the dataframe
leads_details_clean = leads_details.copy()
interactions_clean = interactions_df.copy()
demos_clean = demos_df.copy()
no_interest_clean = no_interest_df.copy()

### Remove the duplicated rows
**Task: Remove duplicate rows**

**Code:**

In [168]:
interactions_clean = interactions_clean.drop_duplicates()

**Test:**

In [169]:
interactions_clean.duplicated().any()

False

### Remove outliers
**Task: Remove outliers in the lead details**

**Code:**

In [170]:
sorted(leads_details.age.unique())

[16, 18, 20, 21, 22, 24, 25, 116, 211]

Any age greater than 25 is an outlier

In [171]:
leads_details_clean = leads_details_clean.query("age <= 25")

**Test:**

In [172]:
leads_details_clean.age.describe()

count    358.000000
mean      20.768156
std        2.852959
min       16.000000
25%       18.000000
50%       21.000000
75%       24.000000
max       25.000000
Name: age, dtype: float64

**Task: Remove outliers in the demo dataset**

**Code:**

In [173]:
## Anything greater than 100% is an outlier
demos_clean = demos_clean.query('watched_percentage<=100')

**Test:**

In [174]:
demos_clean.watched_percentage.max()

100

**Task: Combining reasons columns into 2 columns, stage dropped out and the reason**

**Code:**

In [175]:
stages = {'reasons_for_not_interested_in_demo':'lead',
          'reasons_for_not_interested_to_consider':'awareness',
          'reasons_for_not_interested_to_convert':'consider'}
# Melts the reasons columns into 2 columns
no_interest_clean = no_interest_clean.melt('lead_id',no_interest_clean.columns[1:4],'stage_dropeed','reason').dropna()
# Makes sure the reasons are unique
no_interest_clean.loc[no_interest_clean.reason == 'Cannot afford',"reason"] = "Can't afford"
# Replaces the melted column names with the actual stages
no_interest_clean['stage'] = no_interest_clean['stage_dropeed'].apply(lambda x: stages[x])

**Test:**

In [176]:
no_interest_clean

lead_id                          stage_dropeed  \
0    USR1001     reasons_for_not_interested_in_demo   
32   USR1048     reasons_for_not_interested_in_demo   
33   USR1049     reasons_for_not_interested_in_demo   
35   USR1054     reasons_for_not_interested_in_demo   
37   USR1056     reasons_for_not_interested_in_demo   
..       ...                                    ...   
819  USR1287  reasons_for_not_interested_to_convert   
823  USR1291  reasons_for_not_interested_to_convert   
832  USR1305  reasons_for_not_interested_to_convert   
864  USR1343  reasons_for_not_interested_to_convert   
869  USR1348  reasons_for_not_interested_to_convert   

                               reason     stage  
0                 No time for student      lead  
32   Student not interested in domain      lead  
33                No time for student      lead  
35              Wants offline classes      lead  
37                       Can't afford      lead  
..                                ...       ...  
819             Wants offline classes  consider  
823             Wants offline classes  consider  
832             Wants offline classes  consider  
864             Wants offline classes  consider  
869             Wants offline classes  consider  

[294 rows x 4 columns]

The dataset has now 3 columns, lead_id, stage_dropped and reason

In [177]:
no_interest_clean['reason'].unique()

array(['No time for student', 'Student not interested in domain',
       'Wants offline classes', "Can't afford", 'Will join in final year'],
      dtype=object)

Unique responses

## Saving the cleaned data

In [178]:
leads_details_clean.to_csv('clean/leads_basic_details_clean.csv',index=False)
no_interest_clean.to_csv('clean/leads_reasons_for_no_interest_clean.csv',index=False)
interactions_clean.to_csv('clean/leads_interaction_details_clean.csv',index=False)
demos_clean.to_csv('clean/leads_demo_watched_details_clean.csv',index=False)